clang/lib/Tooling/DependencyScanning/ModuleDepCollector.cpp

```diff
@@ -739,6 +748,12 @@ ModuleDepCollectorPP::handleTopLevelModule(const Module *M) {
   MDC.ScanInstance.getASTReader()->visitInputFileInfos(
       *MF, /*IncludeSystem=*/true,
       [&](const serialization::InputFileInfo &IFI, bool IsSystem) {
+        if (MD.IsInSysroot) {
+          auto FullFilePath = ASTReader::ResolveImportedPath(
+              PathBuf, IFI.UnresolvedImportedFilename, MF->BaseDirectory);
+          MD.IsInSysroot = FullFilePath->starts_with(CurrSysroot);
```

Comment:
Minor nit: string prefix is not correct for paths, it also needs to check that the next character in `FullFilePath` is a separator.  I'm surprised I can't find an existing utility for this in Path.h, but maybe we should add one.

Reply:
There is one private to `Path.cpp` I could expose that?

Reply:
Do you mean `starts_with`? That one looks like it has the same issue 😬 In fact I see in the `replace_path_prefix` documentation it claims the following behaviour
```
/oldfoo, /old, /new => /oldfoo
```

but the unit test for this function actually has an equivalent case and the behaviour is different `/oldnew/foo, /old, /new => /newnew/foo` specifically because `starts_with` is wrong.

Reply:
There's also a swift version that attempts to do the same. https://github.com/swiftlang/swift/blob/b4a2908fb5d7ae5b19d1f4a7f0ec9636f7c729c9/lib/Basic/StringExtras.cpp#L1459
It isolates & moves forward by separator, so I can try to apply this here (and double check it resolves the examples you point out).

Reply:
I modified the test case to handle a case like you described. https://github.com/llvm/llvm-project/pull/130634/files#diff-988370ab7ec8ac1b26c1fa07bda79906b52d097324dac9a9d04774debf330a6cR70